### PR TITLE
[quick video overview]  Mark W&B run as failed when `WandbLogger.finalize("failed")` is called

### LINKED ISSUE #21787: WandbLogger marks crashed runs as "Finished" instead of "Failed" in W&B sweeps
### Bug description

When using W&B sweeps, a run that crashes via `sys.exit(0)` — or any other mechanism that causes Python to exit with code 0 — is incorrectly marked as **"Finished"** (green) in the W&B UI instead of **"Failed"**. This silently corrupts sweep results because a crashed run appears to have completed successfully.

`WandbLogger.finalize("failed")` previously did nothing (early return). Python's atexit handler then calls `wandb.finish()` with the process exit code, and since `sys.exit(0)` produces exit code 0, W&B interprets it as success.

**Expected:** The run is marked as "Failed".
**Actual:** The run is marked as "Finished".

I'll have a PR in a moment.

### What version are you seeing the problem on?

master

### Reproduced in studio

_No response_

### How to reproduce the bug

```python
Save the following as `crash_test.py`:


import sys
import torch
from lightning.pytorch import LightningModule, Trainer
from lightning.pytorch.loggers import WandbLogger
from torch.utils.data import DataLoader, TensorDataset


class CrashingModel(LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(1, 1)

    def training_step(self, batch, batch_idx):
        x, y = batch
        loss = torch.nn.functional.mse_loss(self.layer(x), y)
        self.log("train/loss", loss)
        if batch_idx >= 2:
            sys.exit(0)
        return loss

    def configure_optimizers(self):
        return torch.optim.Adam(self.parameters())


trainer = Trainer(
    max_epochs=3,
    logger=WandbLogger(project="my_project"),
)
trainer.fit(CrashingModel(), DataLoader(TensorDataset(torch.randn(100, 1), torch.randn(100, 1))))


Save the following as `crash_test_sweep.yaml`:


program: crash_test.py
project: my_project
method: random
run_cap: 1
parameters:
  learning_rate:
    values: [0.001, 0.01, 0.1]


Then run:


wandb sweep crash_test_sweep.yaml
wandb agent <sweep-id>


Observe that the run in the W&B sweep dashboard is marked as "Finished" (green) instead of "Failed".
```

### Error messages and logs

```
# Error messages and logs here please
```

### Environment

<details>
  <summary>Current environment</summary>

* CUDA:
	- GPU:
		- NVIDIA GeForce RTX 3090
		- NVIDIA GeForce RTX 3050
	- available:         True
	- version:           12.8
* Lightning:
	- lightning:         2.6.2
	- lightning-utilities: 0.15.3
	- pytorch-lightning: 2.6.2
	- torch:             2.9.1
	- torchmetrics:      1.8.2
	- torchvision:       0.24.1
* System:
	- OS:                Linux
	- architecture:
		- 64bit
		- ELF
	- processor:         x86_64
	- python:            3.10.19
	- release:           6.6.114.1-microsoft-standard-WSL2

</details>

### More info

_No response_

cc @ethanwharris @lantiga @morganmcg1 @borisdayma @scottire @parambharat

### PR DESCRIPTION
**Pull request by pupeno**
## What does this PR do?

Fixes: #21787

I made a quick video overview that might save you some time during review:: https://youtu.be/AVeU2xPCRkg

Fixes `WandbLogger` incorrectly marking a failed run as "Finished" (success) in W&B when training exits with code 0 (e.g. via `sys.exit(0)`).

**Root cause:** `WandbLogger.finalize("failed")` previously returned early without doing anything. When the training process then exits with code 0, Python's atexit handler calls `wandb.finish()` with `exit_code=0`, which W&B interprets as a successful run. This silently corrupts W&B sweep results — a crashed run appears green in the sweep dashboard.

**Fix:** `finalize("failed")` now explicitly calls `experiment.finish(exit_code=1)`, marking the run as failed before atexit can override it with a misleading exit code.

Sample code can be found on https://github.com/pupeno/pytorch-lightning-workspace/tree/main/21788_wandb_finalize_failed_status

<details>
  <summary><b>Before submitting</b></summary>

- [ ] Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

### CODE REVIEW

**Comment by pupeno** [comment]
I updated this to master and resolved the merge conflict.
